python/tvm/relax/backend/contrib/cutlass.py

```diff
@@ -17,84 +17,141 @@
 
 """Pattern table for CUTLASS backend"""
 
-from tvm.relax import transform
+from typing import Mapping, Optional, Tuple
+
+import tvm
+from tvm.contrib.cutlass.build import is_valid_for_cutlass_matmul
+from tvm.relax import Call, Expr, ShapeExpr, transform
+from tvm.relax.dpl import DFPattern
 
 from ..pattern_registry import get_patterns_with_prefix, register_patterns
 from ..patterns import make_fused_bias_activation_pattern, make_matmul_pattern
 
+
+def _get_static_shape(shape: ShapeExpr) -> Optional[Tuple[int]]:
+    result = []
+    for dim in shape.values:
+        if isinstance(dim, tvm.tir.expr.IntImm):
+            result.append(int(dim))
+        else:
+            return None
+    return result
+
+
+def _is_supported_dtype(lhs_dtype, rhs_dtype):
+    """Check if dtypes in the given workload are supported by CUTLASS."""
+    return (
+        (lhs_dtype == "float16" and rhs_dtype == "float16")
+        or (lhs_dtype == "float32" and rhs_dtype == "float32")
+        or (lhs_dtype in ("int8", "uint8") and rhs_dtype in ("int8", "uint8"))
+    )
+
+
+def _check_matmul(
+    match_result: Mapping[DFPattern, Expr],
+    _: Expr,
+) -> bool:
+    matmul_call: Call = None
+    for _, expr in match_result.items():
+        if isinstance(expr, Call) and expr.op.name == "relax.matmul":
```

Comment:
```suggestion
        if isinstance(expr, Call) and isinstance(expr.op, tvm.ir.Op) and expr.op.name == "relax.matmul":
```
it's possible it's calling a PrimFunc, it will be a global var

Reply:
The problem here is it will iterate over all nodes in the pattern, including wildcard that matches anything. This can actually be a problem when there are two consecutive matmuls, if the first matmul is not offloaded (e.g doesn’t pass check func), and when we are checking the match for the second matmul, the match result will have two matmuls because the wildcard matches the first matmul

Reply:
I will fix this in my follow up PR about dynamic shape support.